File: klinara-ios/klinara-ios/Services/Catalog/CatalogService.swift
import Foundation

/// Hizmet kataloğu uçlarının istemci sözleşmesi (Batch 2.1).
///
/// Ekranlar yalnız bu protokole konuşur; arkasında ``LiveCatalogService`` ya da
/// ``MockCatalogService`` durabilir.
protocol CatalogService: Sendable {

    /// `GET /service-categories`
    func categories() async throws -> [ServiceCategory]

    /// `POST /service-categories`
    func createCategory(_ input: CreateServiceCategoryInput) async throws -> ServiceCategory

    /// `PATCH /service-categories/:id`
    func updateCategory(id: String, _ input: UpdateServiceCategoryInput) async throws -> ServiceCategory

    /// `DELETE /service-categories/:id` — **siler değil, pasife alır**.
    func deactivateCategory(id: String) async throws -> ServiceCategory

    /// `GET /services`
    func services() async throws -> [ClinicService]

    /// `GET /services/:id`
    func service(id: String) async throws -> ClinicService

    /// `POST /services`
    func createService(_ input: CreateServiceInput) async throws -> ClinicService

    /// `PATCH /services/:id`
    func updateService(id: String, _ input: UpdateServiceInput) async throws -> ClinicService

    /// `DELETE /services/:id` — **siler değil, pasife alır**.
    func deactivateService(id: String) async throws -> ClinicService
}

struct LiveCatalogService: CatalogService {

    private let client: APIClient

    init(client: APIClient) {
        self.client = client
    }

    func categories() async throws -> [ServiceCategory] {
        let response: ListEnvelope<ServiceCategory> = try await client.send(
            APIRequest.get("service-categories")
        )
        return response.data
    }

    func createCategory(_ input: CreateServiceCategoryInput) async throws -> ServiceCategory {
        try await client.send(APIRequest.post("service-categories", body: input))
    }

    func updateCategory(
        id: String,
        _ input: UpdateServiceCategoryInput
    ) async throws -> ServiceCategory {
        try await client.send(APIRequest.patch("service-categories/\(id)", body: input))
    }

    func deactivateCategory(id: String) async throws -> ServiceCategory {
        try await client.send(APIRequest.delete("service-categories/\(id)"))
    }

    func services() async throws -> [ClinicService] {
        let response: ListEnvelope<ClinicService> = try await client.send(APIRequest.get("services"))
        return response.data
    }

    func service(id: String) async throws -> ClinicService {
        try await client.send(APIRequest.get("services/\(id)"))
    }

    func createService(_ input: CreateServiceInput) async throws -> ClinicService {
        try await client.send(APIRequest.post("services", body: input))
    }

    func updateService(id: String, _ input: UpdateServiceInput) async throws -> ClinicService {
        try await client.send(APIRequest.patch("services/\(id)", body: input))
    }

    func deactivateService(id: String) async throws -> ClinicService {
        try await client.send(APIRequest.delete("services/\(id)"))
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/catalog/CatalogService.kt
package com.klinara.android.services.catalog

/**
 * Hizmet kataloğu — `service:read` okur, `service:write` yazar.
 *
 * Okuma yarısının bir metodu A3.4'e alınmıştı (randevu oluşturmak hizmet seçmeyi
 * gerektiriyor); yazma yarısı ve kategoriler A7.1'de geldi.
 *
 * **Silme yok.** İki `deactivate*` metodu sunucuda `DELETE` ama kaydı silmez, pasife alır
 * ve güncel kaydı döner (204 değil): geçmiş randevular ve paketler hizmete bağlı kalmalı.
 * Aktif hizmeti olan kategori pasife alınamaz — 409 `CONFLICT`, metni sunucudan.
 */
interface CatalogService {
    /** `GET service-categories` — sıralı (`sortOrder`, sonra ad); pasifler dahil. */
    suspend fun categories(): List<ServiceCategory>

    /** `POST service-categories`. Slug çakışması 409. */
    suspend fun createCategory(input: CreateServiceCategoryInput): ServiceCategory

    /** `PATCH service-categories/:id`. */
    suspend fun updateCategory(
        id: String,
        input: UpdateServiceCategoryInput,
    ): ServiceCategory

    /** `DELETE service-categories/:id` — pasife alır, kaydı döner. */
    suspend fun deactivateCategory(id: String): ServiceCategory

    /** `GET services` — kiracının tüm hizmetleri (pasifler dahil; süzme istemcide). */
    suspend fun services(): List<ClinicService>

    /** `GET services/:id`. */
    suspend fun service(id: String): ClinicService

    /** `POST services`. */
    suspend fun createService(input: CreateServiceInput): ClinicService

    /** `PATCH services/:id`. `branchOverrides` verilirse listeyi tamamen değiştirir. */
    suspend fun updateService(
        id: String,
        input: UpdateServiceInput,
    ): ClinicService

    /** `DELETE services/:id` — pasife alır, kaydı döner. */
    suspend fun deactivateService(id: String): ClinicService
}
